File: BlogApp-RN/src/features/authentication/presentation/state/authState.ts
class AuthState {
  constructor(private user: User, public token: Token) {}
  public get signedIn(): boolean {
    return (
      this.user.id !== undefined &&
      this.token.value !== undefined &&
      this.token.expires !== undefined &&
      new Date().getTime() < this.token.expires
    );
  }

  public get email(): string {
    return this.user.email ?? "";
  }

  public get fullName(): string {
    return `${this.user.firstName} ${this.user.lastName}`;
  }
  public get isStudent(): boolean {
    return this.user.role === "student";
  }
  public get isParent(): boolean {
    return this.user.role === "parent";
  }
  public get isInstitution(): boolean {
    return this.user.role === "institution";
  }
  public get isAdmin(): boolean {
    return this.user.role === "admin";
  }

  public copyWith: (params: AuthCopyParams) => AuthState = ({
    user,
    token,
  }) => {
    return new AuthState(user ?? this.user, token ?? this.token);
  };
}

interface AuthCopyParams {
  user?: User;
  token?: Token;
}

interface Token {
  value?: string;
  expires?: number;
}

export interface User {
  id?: string;
  email?: string;
  firstName?: string;
  lastName?: string;
  birthDate?: Date;
  role?: UserRole;
}

type UserRole = "student" | "parent" | "institution" | "admin" | undefined;

export default AuthState;

File: blogapp_flutter/lib/features/authentication/presentation/state/auth_state.dart
import 'package:blogapp_flutter/features/authentication/typedefs/token.dart';
import 'package:blogapp_flutter/features/authentication/typedefs/user.dart';
import 'package:blogapp_flutter/features/authentication/typedefs/user_role.dart';
import 'package:equatable/equatable.dart';
class AuthState extends Equatable {
  final User? user;
  final Token? token;

  const AuthState({
    this.user,
    this.token,
  });

  factory AuthState.initial() {
    return const AuthState(
      user: null,
      token: null,
    );
  }

  bool get signedIn =>
      user?.id != null && token != null && DateTime.now().millisecondsSinceEpoch < token!.expires;

  String get email => user?.email ?? '';

  String get fullName => '${user?.firstName} ${user?.lastName}';
  UserRole? get role => user?.role;
  bool get isStudent => role == UserRole.student;
  bool get isParent => role == UserRole.parent;
  bool get isInstitution => role == UserRole.institution;
  bool get isAdmin => role == UserRole.admin;


  AuthState copyWith({
    User? user,
    Token? token,
  }) {
    return AuthState(
      user: user ?? this.user,
      token: token ?? this.token,
    );
  }

  @override
  List<Object?> get props => [user, token];
}
